base/compiler/ssair/inlining.jl

```diff
@@ -808,155 +820,179 @@ function is_valid_type_for_apply_rewrite(@nospecialize(typ), sv)
+end
 
-        # Special handling for Core.invoke and Core._apply, which can follow the normal inliner
-        # logic with modified inlining target
-        isinvoke = false
+# Handles all analysis and inlining of intrinsics and builtins. In particular,
+# this method does not access the method table or otherwise process generic
+# functions.
+function process_simple!(ir::IRCode, idx::Int, params::Params)
+    stmt = ir.stmts[idx]
+    if isexpr(stmt, :splatnew)
+        inline_splatnew!(ir, idx)
+        return nothing
+    end
 
-        if f !== Core.invoke && (isa(f, IntrinsicFunction) || ft ⊑ IntrinsicFunction || isa(f, Builtin) || ft ⊑ Builtin)
-            # TODO: this test is wrong if we start to handle Unions of function types later
-            continue
-        end
+    isexpr(stmt, :call) || return nothing
 
-        # Handle invoke
-        invoke_data = nothing
-        if f === Core.invoke && length(atypes) >= 3
-            res = compute_invoke_data(atypes, stmt.args, sv)
-            res === nothing && continue
-            (f, ft, atypes, argexprs, invoke_data) = res
-        end
-        isinvoke = (invoke_data !== nothing)
+    sig = call_sig(ir, stmt)
+    sig === nothing && return nothing
 
-        atype = argtypes_to_type(atypes)
+    # Handle _apply
+    sig = inline_apply!(ir, idx, sig, params)
+    sig === nothing && return nothing
 
-        # In :invoke, make sure that the arguments we're passing are a subtype of the
-        # signature we're invoking.
-        (invoke_data === nothing || atype <: invoke_data.types0) || continue
+    # Check if we match any of the early inliners
+    calltype = ir.types[idx]
+    res = early_inline_special_case(ir, sig, stmt, params, calltype)
+    if res !== nothing
+        ir.stmts[idx] = res
+        return nothing
+    end
 
-        # Bail out here if inlining is disabled
-        sv.params.inlining || continue
+    # Bail out here if inlining is disabled
+    params.inlining || return nothing
+
+    # Handle invoke
+    invoke_data = nothing
+    if sig.f === Core.invoke && length(sig.atypes) >= 3
+        res = compute_invoke_data(sig.atypes, params)
+        res === nothing && return nothing
+        (sig, invoke_data) = res
+    elseif is_builtin(sig)
+        # No inlining for builtins (other than what was previously handled)
+        return nothing
+    end
 
-        # Special case inliners for regular functions
-        if late_inline_special_case!(ir, idx, stmt, atypes, f, ft) || is_return_type(f)
-            continue
-        end
+    sig = with_atype(sig)
+
+    # In :invoke, make sure that the arguments we're passing are a subtype of the
+    # signature we're invoking.
+    (invoke_data === nothing || sig.atype <: invoke_data.types0) || return nothing
+
+    # Special case inliners for regular functions
+    if late_inline_special_case!(ir, sig, idx, stmt) || is_return_type(sig.f)
+        return nothing
+    end
+    return (sig, invoke_data)
```

Comment:
perhaps should be an `svec`

Reply:
This is the same signature as the other one above. I'll just leave it as a tuple.